0000-new_pulse_drawer.md

```diff
@@ -0,0 +1,274 @@
+  pass
+```
+
+this is the callback function to extract a set of pulse names (sometime formatting names by request) and the coordinate to draw of each pulse in the `EventsOutputChannels`.
+Both of the arguments and the returned values don't contain the `matplotlib` object, this function can be tested by current testing framework (currently this is written as a part of drawer and there is no way to test without importing `matplotlib`).
+We can prepare a set of such callback function in `utils.py`. Truncation and scaling can be implemented as a part of `utils.py`.
+
+The `schedule_drawer_core` execute required callback functions based on the style sheet and generate a set of objects drawn by backends (this is usually `matplotlib`).
+Finally `schedule_drawer_core` returns `matplotlib.Figure` object if `matplotlib.Axes` objects are not given.
+
+The `pulse_drawer_core` has much simple structure because it just need to call `.samples` method of given `SamplePulse` instance and draw the returned `ndarray`.
+
+The entire work can be separated into several small PRs.
+
+### Phase1: Update `EventsOutputChannels`
+This is the fundamental element of the schedule drawer and thus this should be implemented first.
+Since the input is existing pulse `Schedule` there is no dependency on another drawer components.
+We can prepare the unittest for this object.
+
+### Phase2: Add utility functions
+Utility functions have dependency on `EventsOutputChannels`.
+These are candidates of callback function we need to implement:
+
++ `get_labels`: return label of each pulse
++ `get_max_values`: return maximum value of each pulse
++ `get_min_values`: return minimum value of each pulse
++ `get_lengths`: return length of each pulse
++ `get_phases`: return phase change/set instruction
++ `get_frequencies`: return frequency change/set instruction
++ `get_snapshots`: return snapshot instruction
++ `scale_waveform`: apply scaling to pulses
++ `truncate_waveform`: truncate long waveform and delay
+
+We can prepare unittest for those functions as well.
+
+### Phase3: Add style sheets
+The style sheet doesn't have any dependency so we can start implementation of them at anytime we want.
+These are candidates of style sheets we need to implement:
+
++ `iqx_debugging`: show every details of the pulse, may return very busy plot
++ `iqx_standard`: show instruction level information to understand the overview of program, small details are ignored
++ `iqx_publication`: show minimum information to make clean plot for publication
+
+### Phase4: Add core function and interface
+With all of above PRs we can start implementing main part of drawer.
+The unittest for `matplotlib.Figure` output is skipped in the current testing framework.
+We need to check the output of figure by our eyes.
+
+## Alternative Approaches
+N/A
+
+## Questions
++ How old drawer can be deprecated? Becase interface will change drastically, I feel introducing new drawer independently to old one is easier.
```

Comment:
So long as calling it without arguments is still possible we can replace the old drawer and break code for such a big change. We can introduce the new drawer slowly with a different function name such as `drawv2` to test over time.